### ISSUE TITLE
Swarm panic on trying to create container through API

### ISSUE THREAD

**Issue by sayash** [docker/swarm#1067]
Reporting a potential bug.

I'm sending a POST request to swarm to create a new container, but the socket closes abruptly for me, and when I check swarm logs, this is what I see:

```
2015/07/16 18:56:26 http: panic serving 10.0.2.2:55145: runtime error: invalid memory address or nil pointer dereference  
goroutine 28024 [running]:  
net/http.func·011()  
        /usr/src/go/src/pkg/net/http/server.go:1100 +0xb7  
runtime.panic(0x963280, 0xcd5d33)  
        /usr/src/go/src/pkg/runtime/panic.c:248 +0x18d  
github.com/docker/swarm/api.postContainersCreate(0xc20808f1c0, 0x7f0c26c88c30, 0xc208425360, 0xc208456f70)  
        /go/src/github.com/docker/swarm/api/handlers.go:342 +0x614  
github.com/docker/swarm/api.func·008(0x7f0c26c88c30, 0xc208425360, 0xc208456f70)  
        /go/src/github.com/docker/swarm/api/primary.go:111 +0x2a0  
net/http.HandlerFunc.ServeHTTP(0xc2080efd60, 0x7f0c26c88c30, 0xc208425360, 0xc208456f70)  
        /usr/src/go/src/pkg/net/http/server.go:1235 +0x40  
github.com/gorilla/mux.(*Router).ServeHTTP(0xc2080df090, 0x7f0c26c88c30, 0xc208425360, 0xc208456f70)  
        /go/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x292  
github.com/docker/swarm/api.(*dispatcher).ServeHTTP(0xc208039ed0, 0x7f0c26c88c30, 0xc208425360, 0xc208456f70)  
        /go/src/github.com/docker/swarm/api/server.go:32 +0x9d  
net/http.serverHandler.ServeHTTP(0xc208106de0, 0x7f0c26c88c30, 0xc208425360, 0xc208456f70)  
        /usr/src/go/src/pkg/net/http/server.go:1673 +0x19f  
net/http.(*conn).serve(0xc2084a0700)  
        /usr/src/go/src/pkg/net/http/server.go:1174 +0xa7e  
created by net/http.(*Server).Serve  
        /usr/src/go/src/pkg/net/http/server.go:1721 +0x313  
```
After I restart swarm, it works fine.

The swarm has been created by docker-machine (v0.3.0), Docker engine v1.7.1, which is running the swarm image id - 446e413bd987.

**Comment by djmaze** [comment]
Same here. Restarting `swarm-agent-master`helped.

docker-machine 0.4.0-dev, Docker engine 1.7.1 running swarm 0.3.0/446e413bd987.

**Comment by abronan** [comment]
Hi @sayash and @djmaze, thanks for reporting. Do you encounter this issue on a fresh swarm or after some time using it? Also, do you run any kind of specific configuration: using Replication on a multi-Master setup, etc.?

**Comment by djmaze** [comment]
Hi @abronan. This was a fresh swarm. Running a container failed on the first try. The agents were started before the master. No special configuration otherwise.

**Comment by sayash** [comment]
Hi, this was on a fresh swarm. Master was started before agents.

**Comment by abronan** [comment]
Can't seem to reproduce using the latest `master` and docker `1.7.1`. Could you provide with more details about the commands you are using? (`docker run` or `docker create` command). Also the inspect output of the container would help (on a single docker daemon).

Are you able to reproduce each time you are running your Swarm or just once in a while?

**Comment by djmaze** [comment]
@abronan: Sorry, can't inspect, as the misbehaving cluster does not exist anymore. I was just running `docker run hello-world` as shown in the documentation, but running other containers failed the same way. 

I should add I created that (3 node) cluster with RancherrOS on the Vultr cloud using a development version of [Docker Machine with Vultr support](https://github.com/docker/machine/pull/958). But as `docker info` showed all the details, I am quite sure swarm had been installed correctly.

I just tried again with a similar setup (Ubuntu on Vultr). Can't reproduce, everything's working from the start.

**Comment by abronan** [comment]
Thanks for the info! It seems that there is an issue between Swarm and RancherOS then.

@sayash Are you using a similar setup? (using RancherOS?)

**Comment by sayash** [comment]
Hi Alexandre - I'm not using RancherOS - just Ubuntu, docker-machine to
spin up VirtualBox machines.

Thanks for your follow up, but unfortunately, I'll have to wait and see
when the bug comes back. While I could create a new swarm and try to
reproduce the bug - but my existing swarm is supporting a few applications
at the moment.

I'll post an update if I'm able to reproduce it again. Let me know what log
data I should look out for.

**Comment by spiderpug** [comment]
I encountered the same thing I guess:

- started a new virtualbox with docker-machine  --swarm --swarm-master and so on ("swarm-master")
- started consul server on a different node
- started swarm-node-1 with docker-machine (with the same swarm discovery settings)
- restarted swarm-node-1 machine
- ran `docker run something`

This is the log output:

    time="2015-08-03T08:48:11Z" level=info msg="Listening for HTTP" addr="0.0.0.0:3376" proto=tcp 
    time="2015-08-03T08:48:41Z" level=error msg="consul: Get http://consul.service.consul:8500/v1/kv/docker/swarm/nodes?recurse=&wait=15000ms: dial tcp: i/o timeout" 
    time="2015-08-03T08:48:41Z" level=error msg="Discovery error: Unexpected watch error" 
    time="2015-08-03T08:49:16Z" level=error msg="consul: Get http://consul.service.consul:8500/v1/kv/docker/swarm/nodes?recurse=&wait=15000ms: dial tcp: lookup consul.service.consul: no such host" 
    time="2015-08-03T08:49:16Z" level=error msg="Discovery error: Unexpected watch error" 
    time="2015-08-03T08:50:06Z" level=error msg="consul: Get http://consul.service.consul:8500/v1/kv/docker/swarm/nodes?recurse=&wait=15000ms: dial tcp: i/o timeout" 
    time="2015-08-03T08:50:06Z" level=error msg="Discovery error: Unexpected watch error" 
    time="2015-08-03T08:50:36Z" level=error msg="consul: Get http://consul.service.consul:8500/v1/kv/docker/swarm/nodes?recurse=&wait=15000ms: dial tcp: lookup consul.service.consul: no such host" 
    time="2015-08-03T08:50:36Z" level=error msg="Discovery error: Unexpected watch error" 
    time="2015-08-03T08:51:26Z" level=error msg="consul: Get http://consul.service.consul:8500/v1/kv/docker/swarm/nodes?recurse=&wait=15000ms: dial tcp: i/o timeout" 
    time="2015-08-03T08:51:26Z" level=error msg="Discovery error: Unexpected watch error" 
    time="2015-08-03T08:51:51Z" level=error msg="consul: Get http://consul.service.consul:8500/v1/kv/docker/swarm/nodes?recurse=&wait=15000ms: dial tcp: lookup consul.service.consul: no such host" 
    time="2015-08-03T08:51:51Z" level=error msg="Discovery error: Unexpected watch error" 
    time="2015-08-03T08:52:11Z" level=info msg="Registered Engine swarm-master at 192.168.99.110:2376" 
    time="2015-08-03T08:52:11Z" level=info msg="Registered Engine swarm-node-1 at 192.168.99.111:2376" 
    time="2015-08-03T08:53:48Z" level=error msg="Flagging engine as dead. Updated state failed: Get https://192.168.99.111:2376/v1.15/containers/json?all=1&size=0: dial tcp 192.168.99.111:2376: no route to host" id="YMEH:GSIM:OYUU:XTUD:YBDH:32DJ:54QB:T7PP:ECPE:RCGP:V33K:SDQV" name=swarm-node-1 
    time="2015-08-03T08:54:18Z" level=info msg="Engine came back to life. Hooray!" id="YMEH:GSIM:OYUU:XTUD:YBDH:32DJ:54QB:T7PP:ECPE:RCGP:V33K:SDQV" name=swarm-node-1 
    2015/08/03 08:57:51 http: panic serving 192.168.99.1:62403: runtime error: invalid memory address or nil pointer dereference
    goroutine 108 [running]:
    net/http.func·011()
      /usr/src/go/src/pkg/net/http/server.go:1100 +0xb7
    runtime.panic(0x963280, 0xcd5d33)
      /usr/src/go/src/pkg/runtime/panic.c:248 +0x18d
    github.com/docker/swarm/api.postContainersCreate(0xc20809f040, 0x7f6948c8d9a8, 0xc208212000, 0xc20818e0d0)
      /go/src/github.com/docker/swarm/api/handlers.go:342 +0x614
    github.com/docker/swarm/api.func·008(0x7f6948c8d9a8, 0xc208212000, 0xc20818e0d0)
      /go/src/github.com/docker/swarm/api/primary.go:111 +0x2a0
    net/http.HandlerFunc.ServeHTTP(0xc2080d4340, 0x7f6948c8d9a8, 0xc208212000, 0xc20818e0d0)
      /usr/src/go/src/pkg/net/http/server.go:1235 +0x40
    github.com/gorilla/mux.(*Router).ServeHTTP(0xc20809d1d0, 0x7f6948c8d9a8, 0xc208212000, 0xc20818e0d0)
      /go/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x292
    github.com/docker/swarm/api.(*dispatcher).ServeHTTP(0xc208039ee0, 0x7f6948c8d9a8, 0xc208212000, 0xc20818e0d0)
      /go/src/github.com/docker/swarm/api/server.go:32 +0x9d
    net/http.serverHandler.ServeHTTP(0xc20811a780, 0x7f6948c8d9a8, 0xc208212000, 0xc20818e0d0)
      /usr/src/go/src/pkg/net/http/server.go:1673 +0x19f
    net/http.(*conn).serve(0xc2081be000)
      /usr/src/go/src/pkg/net/http/server.go:1174 +0xa7e
    created by net/http.(*Server).Serve
      /usr/src/go/src/pkg/net/http/server.go:1721 +0x313
    2015/08/03 08:59:08 http: panic serving 192.168.99.1:62499: runtime error: invalid memory address or nil pointer dereference
    goroutine 109 [running]:
    net/http.func·011()
      /usr/src/go/src/pkg/net/http/server.go:1100 +0xb7
    runtime.panic(0x963280, 0xcd5d33)
      /usr/src/go/src/pkg/runtime/panic.c:248 +0x18d
    github.com/docker/swarm/api.postContainersCreate(0xc20809f040, 0x7f6948c8d9a8, 0xc208212280, 0xc20818e000)
      /go/src/github.com/docker/swarm/api/handlers.go:342 +0x614
    github.com/docker/swarm/api.func·008(0x7f6948c8d9a8, 0xc208212280, 0xc20818e000)
      /go/src/github.com/docker/swarm/api/primary.go:111 +0x2a0
    net/http.HandlerFunc.ServeHTTP(0xc2080d4340, 0x7f6948c8d9a8, 0xc208212280, 0xc20818e000)
      /usr/src/go/src/pkg/net/http/server.go:1235 +0x40
    github.com/gorilla/mux.(*Router).ServeHTTP(0xc20809d1d0, 0x7f6948c8d9a8, 0xc208212280, 0xc20818e000)
      /go/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x292
    github.com/docker/swarm/api.(*dispatcher).ServeHTTP(0xc208039ee0, 0x7f6948c8d9a8, 0xc208212280, 0xc20818e000)
      /go/src/github.com/docker/swarm/api/server.go:32 +0x9d
    net/http.serverHandler.ServeHTTP(0xc20811a780, 0x7f6948c8d9a8, 0xc208212280, 0xc20818e000)
      /usr/src/go/src/pkg/net/http/server.go:1673 +0x19f
    net/http.(*conn).serve(0xc2081bed00)
      /usr/src/go/src/pkg/net/http/server.go:1174 +0xa7e
    created by net/http.(*Server).Serve
      /usr/src/go/src/pkg/net/http/server.go:1721 +0x313

**Comment by abronan** [comment]
Thanks for the feedback @sayash and @spiderpug.

Trying to find some common ground, @sayash did you restart one of the Swarm Agent somehow during the process? (which can explain why you cannot reproduce now if you have a fresh cluster)

There are a few Open Issues in `docker/machine` with the Swarm Agent restart process (see docker/machine#1518, docker/machine#806).

Seems like what happened is that:
- You create the cluster with `docker-machine`
- Swarm Manager registers the machine
- We restart one of the Swarm agents (the IP changes and/or the certs are not valid anymore)
- Swarm Manager still references the old machine and old TLS config which might somehow lead to the panic on the `postContainerCreate` with `dockerclient`.

This also explains why restarting the Swarm Manager seems to solve the issue. I will investigate and try to reproduce using the same process.

/cc @ehazlett @nathanleclaire

**Comment by mountkin** [comment]
The `cluster.createContainer` should not return nil on this line https://github.com/docker/swarm/blob/master/cluster/swarm/cluster.go#L141.
I opened a PR https://github.com/docker/swarm/pull/1115 to raise an error when the scheduled engine get removed.
But I'm not sure whether the PR can fix this issue.

**Comment by aluzzardi** [comment]
Duplicate of #670

**Comment by abronan** [comment]
I really don't think it's a duplicate so I'll reopen until we find exactly what it is. AFAIK there are still issues on `postContainersCreate` and `refreshContainers`.

**Comment by agsmorodin** [comment]
I still see the exceptions in logs when I am trying to create the container, docker version:
swarm version 0.5.0-dev (76591d2

```
2015/09/25 10:09:05 http: panic serving 10.10.92.93:48345: runtime error: index out of range
goroutine 143456 [running]:
net/http.(*conn).serve.func1(0xc8b29f3600, 0x7f54942e5f00, 0xc8b1f49340)
	/usr/local/go/src/net/http/server.go:1287 +0xb5
github.com/docker/swarm/scheduler/filter.(*AffinityFilter).Filter(0xdd3d38, 0xc8b384e380, 0xc8b3430540, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/swarm/scheduler/filter/affinity.go:37 +0x13aa
github.com/docker/swarm/scheduler/filter.ApplyFilters(0xc8200cb500, 0x5, 0x8, 0xc8b384e380, 0xc8b3430540, 0x2, 0x2, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/docker/swarm/scheduler/filter/filter.go:61 +0x12d
github.com/docker/swarm/scheduler.(*Scheduler).SelectNodeForContainer(0xc8200d91a0, 0xc8b3430540, 0x2, 0x2, 0xc8b384e380, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/swarm/scheduler/scheduler.go:31 +0x99
github.com/docker/swarm/cluster/swarm.(*Cluster).createContainer(0xc8200cdaa0, 0xc8b384e380, 0xc8b2408423, 0x14, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/swarm/cluster/swarm/cluster.go:119 +0x4fb
github.com/docker/swarm/cluster/swarm.(*Cluster).CreateContainer(0xc8200cdaa0, 0xc8b384e380, 0xc8b2408423, 0x14, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/swarm/cluster/swarm/cluster.go:89 +0x6e
github.com/docker/swarm/api.postContainersCreate(0xc8200db6c0, 0x7f549432b5b0, 0xc8b1347550, 0xc8b4c13180)
	/go/src/github.com/docker/swarm/api/handlers.go:379 +0x295
github.com/docker/swarm/api.NewPrimary.func1(0x7f549432b5b0, 0xc8b1347550, 0xc8b4c13180)
	/go/src/github.com/docker/swarm/api/primary.go:120 +0x300
net/http.HandlerFunc.ServeHTTP(0xc820219240, 0x7f549432b5b0, 0xc8b1347550, 0xc8b4c13180)
	/usr/local/go/src/net/http/server.go:1422 +0x3a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc8200d78b0, 0x7f549432b5b0, 0xc8b1347550, 0xc8b4c13180)
	/go/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x29e
github.com/docker/swarm/api.(*dispatcher).ServeHTTP(0xc8200a7c10, 0x7f549432b5b0, 0xc8b1347550, 0xc8b4c13180)
	/go/src/github.com/docker/swarm/api/server.go:32 +0x9f
net/http.serverHandler.ServeHTTP(0xc820288b40, 0x7f549432b5b0, 0xc8b1347550, 0xc8b4c13180)
	/usr/local/go/src/net/http/server.go:1862 +0x19e
net/http.(*conn).serve(0xc8b29f3600)
	/usr/local/go/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1910 +0x3f6
2015/09/25 10:12:59 http: panic serving 10.10.92.93:48346: runtime error: index out of range
goroutine 143680 [running]:
net/http.(*conn).serve.func1(0xc8b98c4370, 0x7f54942e5f00, 0xc8b2a2b340)
	/usr/local/go/src/net/http/server.go:1287 +0xb5
github.com/docker/swarm/api.getContainersJSON(0xc8200db6c0, 0x7f549432b5b0, 0xc8bbf9dd90, 0xc8ae2dca80)
	/go/src/github.com/docker/swarm/api/handlers.go:242 +0x1adc
github.com/docker/swarm/api.NewPrimary.func1(0x7f549432b5b0, 0xc8bbf9dd90, 0xc8ae2dca80)
	/go/src/github.com/docker/swarm/api/primary.go:120 +0x300
net/http.HandlerFunc.ServeHTTP(0xc8200ebe00, 0x7f549432b5b0, 0xc8bbf9dd90, 0xc8ae2dca80)
	/usr/local/go/src/net/http/server.go:1422 +0x3a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc8200d78b0, 0x7f549432b5b0, 0xc8bbf9dd90, 0xc8ae2dca80)
	/go/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x29e
github.com/docker/swarm/api.(*dispatcher).ServeHTTP(0xc8200a7c10, 0x7f549432b5b0, 0xc8bbf9dd90, 0xc8ae2dca80)
	/go/src/github.com/docker/swarm/api/server.go:32 +0x9f
net/http.serverHandler.ServeHTTP(0xc820288b40, 0x7f549432b5b0, 0xc8bbf9dd90, 0xc8ae2dca80)
	/usr/local/go/src/net/http/server.go:1862 +0x19e
net/http.(*conn).serve(0xc8b98c4370)
	/usr/local/go/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1910 +0x3f6
2015/09/25 10:24:37 http: panic serving 10.10.93.62:49187: runtime error: index out of range
goroutine 144317 [running]:
net/http.(*conn).serve.func1(0xc8ad2791e0, 0x7f54942e5f00, 0xc8ab586840)
	/usr/local/go/src/net/http/server.go:1287 +0xb5
github.com/docker/swarm/api.getContainersJSON(0xc8200db6c0, 0x7f549432b5b0, 0xc8b0189a20, 0xc82016b5e0)
	/go/src/github.com/docker/swarm/api/handlers.go:242 +0x1adc
github.com/docker/swarm/api.NewPrimary.func1(0x7f549432b5b0, 0xc8b0189a20, 0xc82016b5e0)
	/go/src/github.com/docker/swarm/api/primary.go:120 +0x300
net/http.HandlerFunc.ServeHTTP(0xc8200ebe00, 0x7f549432b5b0, 0xc8b0189a20, 0xc82016b5e0)
	/usr/local/go/src/net/http/server.go:1422 +0x3a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc8200d78b0, 0x7f549432b5b0, 0xc8b0189a20, 0xc82016b5e0)
	/go/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x29e
github.com/docker/swarm/api.(*dispatcher).ServeHTTP(0xc8200a7c10, 0x7f549432b5b0, 0xc8b0189a20, 0xc82016b5e0)
	/go/src/github.com/docker/swarm/api/server.go:32 +0x9f
net/http.serverHandler.ServeHTTP(0xc820288b40, 0x7f549432b5b0, 0xc8b0189a20, 0xc82016b5e0)
	/usr/local/go/src/net/http/server.go:1862 +0x19e
net/http.(*conn).serve(0xc8ad2791e0)
	/usr/local/go/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1910 +0x3f6
2015/09/25 10:24:40 http: panic serving 10.10.93.62:49188: runtime error: index out of range
goroutine 144326 [running]:
net/http.(*conn).serve.func1(0xc8b0189ad0, 0x7f54942e5f00, 0xc8b38c9080)
	/usr/local/go/src/net/http/server.go:1287 +0xb5
github.com/docker/swarm/api.getContainersJSON(0xc8200db6c0, 0x7f549432b5b0, 0xc8b07193f0, 0xc8af98a1c0)
	/go/src/github.com/docker/swarm/api/handlers.go:242 +0x1adc
github.com/docker/swarm/api.NewPrimary.func1(0x7f549432b5b0, 0xc8b07193f0, 0xc8af98a1c0)
	/go/src/github.com/docker/swarm/api/primary.go:120 +0x300
net/http.HandlerFunc.ServeHTTP(0xc8200ebe00, 0x7f549432b5b0, 0xc8b07193f0, 0xc8af98a1c0)
	/usr/local/go/src/net/http/server.go:1422 +0x3a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc8200d78b0, 0x7f549432b5b0, 0xc8b07193f0, 0xc8af98a1c0)
	/go/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x29e
github.com/docker/swarm/api.(*dispatcher).ServeHTTP(0xc8200a7c10, 0x7f549432b5b0, 0xc8b07193f0, 0xc8af98a1c0)
	/go/src/github.com/docker/swarm/api/server.go:32 +0x9f
net/http.serverHandler.ServeHTTP(0xc820288b40, 0x7f549432b5b0, 0xc8b07193f0, 0xc8af98a1c0)
	/usr/local/go/src/net/http/server.go:1862 +0x19e
net/http.(*conn).serve(0xc8b0189ad0)
	/usr/local/go/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1910 +0x3f6

```

**Comment by abronan** [comment]
@agsmorodin This is a different issue, take a look at #985. Updating docker to the latest should do the trick.

**Comment by abronan** [comment]
This should be fixed as of swarm `v1.0.1`. Closing.
